### DISCUSSION TITLE
Unable to install Python3.13

### DISCUSSION THREAD

**Discussion by Azulath** [Homebrew/discussions#5805] [Everyday usage]
### Output of `brew config`

```shell
❯ brew config
HOMEBREW_VERSION: 4.4.11
ORIGIN: https://github.com/Homebrew/brew
HEAD: ad356d365837a691a52dfec32b3fb5f282f3aefc
Last commit: 4 days ago
Branch: stable
Core tap JSON: 13 Dec 20:12 UTC
Core cask tap JSON: 13 Dec 20:12 UTC
HOMEBREW_PREFIX: /Users/azulath/.homebrew
HOMEBREW_REPOSITORY: /Users/azulath/.homebrew
HOMEBREW_CELLAR: /Users/azulath/.homebrew/Cellar
HOMEBREW_CASK_OPTS: ["--appdir=/Users/azulath/Applications"]
HOMEBREW_EDITOR: /usr/bin/vim
HOMEBREW_MAKE_JOBS: 12
Homebrew Ruby: 3.3.6 => /Users/azulath/.homebrew/Library/Homebrew/vendor/portable-ruby/3.3.6/bin/ruby
CPU: dodeca-core 64-bit arm_lobos
Clang: 16.0.0 build 1600
Git: 2.39.5 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 14.7.2-arm64
CLT: 16.2.0.0.1.1733547573
Xcode: 16.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
❯ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: No Cask quarantine support available: unknown reason.

Warning: Your Homebrew's prefix is not /opt/homebrew.

Many of Homebrew's bottles (binary packages) can only be used with the default prefix.
Consider uninstalling Homebrew and reinstalling into the default prefix.
It is expected behaviour that some formulae will fail to build in this unsupported configuration.
It is expected behaviour that Homebrew will be buggy and slow.
Do not create any issues about this on Homebrew's GitHub repositories.
Do not create any issues even if you think this message is unrelated.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

### Description of issue

When I try to install `Python@3.13` like shown below I get the following error:

```sh
  ~ ❯ brew install python                                                                                                                                                3s  09:21:23 PM
==> Downloading https://formulae.brew.sh/api/formula.jws.json
###################################################################################################################################################################################### 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
###################################################################################################################################################################################### 100.0%
==> Fetching python@3.13
Warning: Building python@3.13 from source as the bottle needs:
- HOMEBREW_CELLAR: /opt/homebrew/Cellar (yours is /Users/azulath/.homebrew/Cellar)
- HOMEBREW_PREFIX: /opt/homebrew (yours is /Users/azulath/.homebrew)
==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-core/b458b354c9ab1d24e0e07e9cfeb26ae4a93cbf3b/Formula/p/python@3.13.rb
###################################################################################################################################################################################### 100.0%
==> Downloading https://raw.githubusercontent.com/Homebrew/formula-patches/8b5bcbb262d1ea4e572bba55043bf7d2341a6821/python/3.13-sysconfig.diff
Already downloaded: /Users/azulath/Library/Caches/Homebrew/downloads/90a805f92fc2334898eb41d25f9b55894305ba85c25fb4020488ec6aa89d4410--3.13-sysconfig.diff
==> Downloading https://files.pythonhosted.org/packages/d5/ae/09427bea9227a33ec834ed5461432752fd5d02b14f93dd68406c91684622/flit_core-3.10.1.tar.gz
Already downloaded: /Users/azulath/Library/Caches/Homebrew/downloads/8c47a1c6fdac341fd6ba8827dabfa3efff834f4a2fcc87c4b0ced64cc0205efa--flit_core-3.10.1.tar.gz
==> Downloading https://files.pythonhosted.org/packages/f4/b1/b422acd212ad7eedddaf7981eee6e5de085154ff726459cf2da7c5a184c1/pip-24.3.1.tar.gz
Already downloaded: /Users/azulath/Library/Caches/Homebrew/downloads/aa552e4585b56ea710159fccadb92eb1c3bd3418bc82bbc0125f0ec73babb823--pip-24.3.1.tar.gz
==> Downloading https://files.pythonhosted.org/packages/43/54/292f26c208734e9a7f067aea4a7e282c080750c4546559b58e2e45413ca0/setuptools-75.6.0.tar.gz
Already downloaded: /Users/azulath/Library/Caches/Homebrew/downloads/703903ddc32daa37a6a497fd460aef21d779a1e435ef2e0181ed99633bea827a--setuptools-75.6.0.tar.gz
==> Downloading https://files.pythonhosted.org/packages/8a/98/2d9906746cdc6a6ef809ae6338005b3f21bb568bea3165cfc6a243fdc25c/wheel-0.45.1.tar.gz
Already downloaded: /Users/azulath/Library/Caches/Homebrew/downloads/2c12eb87bb57b7b29ad1d471909ef6a4e49ed7a89fc6f75dfcc23cb9431321a6--wheel-0.45.1.tar.gz
==> Downloading https://www.python.org/ftp/python/3.13.1/Python-3.13.1.tgz
Already downloaded: /Users/azulath/Library/Caches/Homebrew/downloads/6259c1814b577b7fc34dee20c2046c8b503e52aa1c2ab4167a5d6c06c79f872a--Python-3.13.1.tgz
==> Patching
==> Applying 3.13-sysconfig.diff
==> ./configure --enable-ipv6 --datarootdir=/Users/azulath/.homebrew/Cellar/python@3.13/3.13.1/share --datadir=/Users/azulath/.homebrew/Cellar/python@3.13/3.13.1/share --without-ensurepip -
==> make
Last 15 lines from /Users/azulath/Library/Logs/Homebrew/python@3.13/02.make:
0:00:32 load avg: 3.92 [39/44] test_str
0:00:33 load avg: 3.92 [40/44] test_struct
0:00:33 load avg: 3.92 [41/44] test_tabnanny
0:00:33 load avg: 3.92 [42/44] test_time
0:00:35 load avg: 3.76 [43/44] test_xml_etree
test test_xml_etree failed
0:00:36 load avg: 3.76 [44/44] test_xml_etree_c -- test_xml_etree failed (4 failures)
test test_xml_etree_c failed
test_xml_etree_c failed (3 failures)

Total duration: 36.6 sec
Total tests: run=9,394 failures=7 skipped=198
Total test files: run=44/44 failed=2
Result: FAILURE
make: *** [profile-run-stamp] Error 2

Do not report this issue to Homebrew/brew or Homebrew/homebrew-core!

These open issues may also help:
segmentation fault in python-tk@3.12 and python-tk@3.13 https://github.com/Homebrew/homebrew-core/issues/199354
Do not report this issue: you are running in an unsupported configuration.
```

Is this a general issue with 3.13, since 3.12 works without any issues? So, should I report this as a bug? Or is there something I am doing wrong?

**Comment by p-linnane** [comment]
The output of your `brew doctor` run covers this. You are using an unsupported configuration, so you have to build Python 3.13 from source. If you install into the default prefix (`/opt/homebrew`) you'll be able to use our build bottles instead of compiling on your machine.

**Comment by Azulath** [comment]
But shouldn't compiling also work? I have been using this path for years without any issues and I was also able to install python on my older macbook.

I mean this is just a compiling software and therefore it shod still work fine.

**Reply by p-linnane** [reply]
No, there's no guarantee it should work for you. Installing in a non-default prefix and building from source is specifically called out as unsupported in our [`documentation`](https://docs.brew.sh/Installation#untar-anywhere-unsupported). Saying it's "just compiling software" is not a very well-informed opinion given how complex it really is, not to mention the additional complexities of distribution.

**Reply by Azulath** [reply]
That's true but to a certain extent there should not be any issues by default.

That being said, I was able to install Python via homebrew just be deleting the `~/Library/Caches/Homebrew` folder. I have no idea if something was changed upstream or if my cache was wonky....

**Reply by p-linnane** [reply]
>That's true but to a certain extent there should not be any issues by default.

Feel free to open PR's proposing the changes you'd like to see. Comments like this on their own are very unhelpful and ignore the massive amount of work it takes to maintain this project.

**Reply by Azulath** [reply]
That's not what I meant. Much rather, I have argued that even though I am not using the default installation path - `/opt/homebrew` - the default assumption should not be that this is the root cause of any issues that arises. Hence, I am only challenging the last sentence of the documentation which states:

> If you decide to use another prefix: don’t open any issues, even if you think they are unrelated to your prefix choice. They will be closed without response.

If the documentation is on Github, I can certainly open a PR and remove this statement, but I assume this is not the intent of the authors/maintainers here.

Additionally, I have not directly created a bug report but rather opened a discussion which to my understanding is allowed. Aside from that, as I have mentioned above, deleting the cache and re-installing it did solve the issue even though I would it would be interesting to know what may have caused this.

**Reply by carlocab** [reply]
> Aside from that, as I have mentioned above, deleting the cache and re-installing it did solve the issue even though I would it would be interesting to know what may have caused this.

Building software from source is incredibly fragile and temperamental -- their results can change, or they can suddenly fail, the moment you look at them wrong. This is why we:
1. provide binary packages;
2. don't provide support for building from source; and,
3. ask users not to file issues for failed source builds.

Deleting the cache and then attempting an install might have helped. Or, it could be a complete red herring. It's impossible to say, given the fragile nature of source builds.

**Reply by hangxie** [reply]
This is more like a bug from upstream that python 3.13 needs newer version of expat, but the code to handle older version of expat does not work as expected, I posted a workground to https://github.com/python/cpython/issues/128838#issuecomment-2614238449, which also links to cython's discussion about expat.

**Reply by branchv** [reply]
FYI, this was fixed in https://github.com/Homebrew/homebrew-core/pull/206568 by depending on brewed `expact` for macOS <15
